src/Broadway/EventStore/MongoDBEventStore.php

```diff
@@ -0,0 +1,367 @@
+        }
+
+        if (count($events) == 1) {
+            try {
+                $this->saveEvent(reset($events));
+            } catch (MongoException $exception) {
+                throw MongoDBEventStoreException::create($exception);
+            } catch (Exception $exception) {
+                throw new MongoDBEventStoreException($exception->getMessage(), 0, $exception);
+            }
+
+            return;
+        }
+
+        try {
+            $this->saveBulkEvents($events);
+        } catch (MongoException $exception) {
+            throw MongoDBEventStoreException::create($exception);
+        } catch (Exception $exception) {
+            throw new MongoDBEventStoreException($exception->getMessage(), 0, $exception);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configureCollection()
+    {
+        $this->eventCollection->ensureIndex(array(
+            self::FIELD_STREAM_ID => 1,
+            self::FIELD_PLAYHEAD  => 1,
+        ), array(
+            'unique' => true,
+            'background' => true,
+        ));
+    }
+
+    /**
+     * @param DomainEventStreamInterface|DomainMessage[] $eventStream
+     * @param string                                     $aggregateId
+     *
+     * @return array
+     */
+    private function extractEvents(DomainEventStreamInterface $eventStream, $aggregateId)
+    {
+        $events = array();
+
+        foreach ($eventStream as $message) {
+            $this->assertIdentifier($aggregateId, $message->getId());
+
+            $events[] = $this->serializeEvent($message);
+        }
+
+        return $events;
+    }
+
+    /**
+     * @param string $expected
+     * @param string $actual
+     */
+    private function assertIdentifier($expected, $actual)
+    {
+        if (0 != strcmp($expected, $actual)) {
+            throw new InvalidIdentifierException(sprintf('Expecting %s, got %s identifier.', $expected, $actual));
+        }
+    }
+
+    /**
+     * @param DomainMessage $message
+     *
+     * @return array
+     */
+    private function serializeEvent(DomainMessage $message)
+    {
+        $data = array(
+            self::FIELD_STREAM_ID   => (string) $message->getId(),
+            self::FIELD_PLAYHEAD    => $message->getPlayhead(),
+            self::FIELD_METADATA    => $this->metadataSerializer->serialize($message->getMetadata()),
+            self::FIELD_PAYLOAD     => $this->payloadSerializer->serialize($message->getPayload()),
+            self::FIELD_RECORDED_ON => $message->getRecordedOn()->toString(),
+            self::FIELD_TYPE        => $message->getType(),
+        );
+
+        // save space
+        if (empty($data[self::FIELD_METADATA])) {
+            unset($data[self::FIELD_METADATA]);
+        }
+
+        return $data;
+    }
+
+    /**
+     * @param array $row
+     *
+     * @return DomainMessage
+     */
+    private function deserializeEvent(array $row)
+    {
+        if (empty($row[self::FIELD_METADATA])) {
```

Comment:
It depends on serialization. Eg, save only payload, as class name is known and never to be changed. Or why save empty payload of metadata, imagine overhead of 100M docs.